### PR TITLE
BioMetadata Protocol

### PR DESCRIPTION
**Pull request by david4096**
Adds the BioSample and Individual objects in coordination with https://github.com/ga4gh/schemas/pull/636 https://github.com/ga4gh/compliance/pull/195.

This allows Calls and Read Groups to be filtered by a specific `bioSampleID`. Both individuals and biosamples can be searched by name, and biosamples can be further filtered by individual ID. The two new endpoints added are `/individuals` and `/biosamples`.

Minimum changes to the variants and reads APIs have been made to support this feature. Providing a bioSampleId is optional. This same pattern could be applied to future objects as well.

Known issues:

In practice one needs an individual ID when adding a biosample record to take complete advantage of the metadata linkages. An example of this has been provided in `prepare_compliance_data`.

A bioSampleId can modify the read group set record at query time, which can be confusing, in the future we may consider a separate endpoint for searching read groups.

### CODE REVIEW

File: ga4gh/backend.py [hunk 1]
```diff
@@ -633,16 +701,18 @@ def callSetsGenerator(self, request):
...
-            return self._singleObjectGenerator(callSet)
+        results = []
+        for obj in variantSet.getCallSets():
+            include = True
+            if request.name:
+                if request.name != obj.getLocalId():
+                    include = False
+            if request.bio_sample_id:
+                if request.bio_sample_id != obj.getBioSampleId():
+                    include = False and include
```
Comment (jeromekelleher) [inline ga4gh/backend.py]: Isn't the ``and include`` redundant here? False && x == False.

**Comment by jeromekelleher** [comment]
This is truly excellent @david4096, nice, simple, straightforward code. I haven't tried it out and played with the new protocol yet, but I'm confident this code will go into master without any problems. Great work.

File: docs/configuration.rst [hunk 2]
```diff
@@ -168,6 +168,50 @@ Adds a reference set used in the 1000 Genomes project using the name
...
+   :func: getRepoManagerParser
+   :prog: ga4gh_repo
+   :path: add-individual
+   :nodefault:
+
+**Examples:**
+
+.. code-block:: bash
+
+    $ ga4gh_repo add-individual registry.db dataset1 HG00096 '{"individualId": "abc"}'
```
Comment (macieksmuga) [inline docs/configuration.rst]: You probably wanted `'{"id": "abc"}'` there

**Comment by macieksmuga** [comment]
+1 modulo documentation quibble. Also, `ga4gh_repo list` doesn't list the added biosamples or individuals (may be nice to add these features). I ran and tested the code with the compliance suite test data - all works as expected.

**Comment by macieksmuga** [comment]
Could we one more +1 for this, to allow merging? @jeromekelleher @diekhans @calbach @dcolligan, anybody else you can think of. Thanks!

**Comment by skeenan** [comment]
@mbaudis

**Comment by mbaudis** [comment]
+1

File: docs/api.rst [hunk 3]
```diff
@@ -25,6 +25,17 @@ Types
 
 .. todo:: Add links to the upstream documentation for the GA4GH types.
 
+
```
Comment (jeromekelleher) [inline docs/api.rst]: Please remove this section --- this will just print out confusing protobuf stuff into the docs.

**Comment by jeromekelleher** [comment]
Great work @david4096, +1 modulo (very) minor issue with the docs.

File: docs/api.rst (hunk 3 above)
Comment (kozbo) [inline docs/api.rst]: @jeromekelleher quick clarification: do you mean the todo section or the following BioMetadata section?

**Pull request merged**